src/main/java/com/marklogic/spark/reader/file/MlcpArchiveFileReader.java

```diff
@@ -19,39 +24,74 @@ class MlcpArchiveFileReader implements PartitionReader<InternalRow> {
 
     private final String path;
     private final ZipInputStream zipInputStream;
-    private ZipEntry currentZipEntry;
+    private final MlcpMetadataConverter mlcpMetadataConverter;
+    private InternalRow nextRowToReturn;
 
     MlcpArchiveFileReader(FilePartition filePartition, FileContext fileContext) {
         this.path = filePartition.getPath();
         this.zipInputStream = new ZipInputStream(fileContext.open(filePartition));
+        this.mlcpMetadataConverter = new MlcpMetadataConverter();
     }
 
+    // TODO Add support for abortOnFailure.
     @Override
     public boolean next() throws IOException {
-        currentZipEntry = FileUtil.findNextFileEntry(zipInputStream);
-        return currentZipEntry != null;
-    }
+        // MLCP always includes a metadata entry, even if the user asks for no metadata. And the metadata entry is
```

Comment:
@anu3990  I think you should be able to use a similar approach for reading an NT archive - though the order is opposite, as our NT archive has the content first, followed by the metadata.

Note the "TODO" as well. To support "abortOnFailure=false", we need to try building the next row in `next()`. If that fails and abortOnFailure=false, we log the error and return false. I still need to add support for that in this class.